macros/src/ast_node.rs

```diff
@@ -0,0 +1,172 @@
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote, ToTokens};
+use std::collections::HashMap;
+use syn::{parse_quote, punctuated::Punctuated, Fields, Ident, ItemEnum, Token};
+
+pub struct State<'a> {
+    enums: &'a [ItemEnum],
+    tree: HashMap<&'a Ident, Vec<&'a Ident>>,
+}
+
+impl<'a> State<'a> {
+    pub fn new(enums: &'a [ItemEnum]) -> Self {
+        Self {
+            enums,
+            tree: enums
+                .iter()
+                .map(|e| (&e.ident, e.variants.iter().map(|v| &v.ident).collect()))
+                .collect(),
+        }
+    }
+
+    pub fn generate_enum_definitions(&self, stream: &mut TokenStream) {
+        for flavor in FLAVORS {
+            for definition in self.enums {
+                let new_def = flavor.transform(definition, |x| self.tree.contains_key(x));
+                new_def.to_tokens(stream);
+            }
+        }
+    }
+
+    pub fn generate_transitive_from_impls(&self, stream: &mut TokenStream) {
+        let root = &self.enums[0].ident;
+
+        for category in &self.tree[root] {
+            for node in &self.tree[category] {
+                if let Some(leaves) = self.tree.get(node) {
+                    // From<KeywordEnum> for NodeEnum
+                    self.emit_from_impl([node, category, root], true, stream);
+
+                    for leaf in leaves {
+                        // From<KeywordTime> for LeafEnum
+                        self.emit_from_impl([leaf, node, category], false, stream);
+                        // From<KeywordTime> for NodeEnum
+                        self.emit_from_impl([leaf, category, root], false, stream);
+                    }
+                } else {
+                    // From<Argument> for NodeEnum>
+                    self.emit_from_impl([node, category, root], false, stream);
+                }
+            }
+        }
+    }
+
+    fn emit_from_impl(
+        &self,
+        [inner, middle, outer]: [&Ident; 3],
+        inner_is_already_enum: bool,
+        stream: &mut TokenStream,
+    ) {
+        for flavor in FLAVORS {
+            let inner_ty = if inner_is_already_enum {
+                flavor.suffix_and_lifetime(inner)
+            } else {
+                flavor.ampersand(inner)
+            };
+            let middle_ty = flavor.suffix(middle);
+            let outer_ty = flavor.suffix_and_lifetime(outer);
+
+            let lifetime = (flavor != Flavor::Val).then_some(quote!(<'n>));
+
+            quote! {
```

Comment:
Unfortunately our build system is somewhat broken, so you'll need to extend this workaround to get reliable rebuilds when using cmake:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 342e6fb55..b332387f3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -56,3 +56,3 @@ add_definitions(-DCMAKE_SOURCE_DIR="${REAL_CMAKE_SOURCE_DIR}")
 # not run cargo, so we need to tell it *something*
-FILE(GLOB sources src/* src/*/* src/*/*/*)
+FILE(GLOB sources Cargo.* src/* src/*/* src/*/*/* macros/Cargo.toml macros/src/*)
```

would be nice to fix this in a general way, or at least detect this problem in CI.
But we don't add macro crates often.

Reply:
Done.

Reply:
Updated to `FILE(GLOB sources Cargo.* src/* src/*/* src/*/*/* macros/* macros/*/*)`.